File: backend/app/services/scrapers/validators/position_title_validator.py
import re

from app.schemas.scraped_job import ScrapedJob
from app.services.scrapers.validators.scraped_job_validator import ScrapedJobValidator


# ==========================================
# Role Targeting (SDE / MLE / Data Science / GenAI-AI Engineering only)
# ==========================================
# Word-boundary phrase patterns -- title must match at least ONE of these.
# This avoids the old bug where a bare "ai" or "data" substring matched
# things like "maintain", "chair", "database administrator", etc.
ROLE_PATTERNS = [
    # Software / SDE
    r"\bsoftware engineer\b", r"\bsoftware developer\b", r"\bsde\b",
    r"\bbackend engineer\b", r"\bback[- ]end engineer\b",
    r"\bfull[- ]stack engineer\b", r"\bapplication engineer\b",

    # ML Engineering
    r"\bmachine learning engineer\b", r"\bml engineer\b", r"\bmle\b",
    r"\bmlops engineer\b",

    # Data Science
    r"\bdata scientist\b", r"\bdata science\b",
    r"\bapplied scientist\b",

    # GenAI / AI Engineering
    r"\bai engineer\b", r"\bgenai engineer\b", r"\bgenerative ai\b",
    r"\bllm engineer\b", r"\bnlp engineer\b", r"\bai\/ml engineer\b",
    r"\bartificial intelligence engineer\b",
]
ROLE_REGEX = re.compile("|".join(ROLE_PATTERNS), re.IGNORECASE)

# Seniority exclusions, matched as whole words/phrases so "head" doesn't
# false-positive on "Headquarters" etc.
SENIORITY_PATTERNS = [
    r"\bsenior\b", r"\bsr\.?\b", r"\blead\b", r"\bstaff\b", r"\bprincipal\b",
    r"\bmanager\b", r"\bdirector\b", r"\bhead of\b", r"\bvp\b",
    r"\bvice president\b", r"\barchitect\b",
]
SENIORITY_REGEX = re.compile("|".join(SENIORITY_PATTERNS), re.IGNORECASE)


class PositionTitleValidator(ScrapedJobValidator):
    def _do_validate(self, job: ScrapedJob) -> bool:
        try:
            if not job.title:
                return False
            if SENIORITY_REGEX.search(job.title):
                return False
            if not ROLE_REGEX.search(job.title):
                return False
        except Exception as e:
            print(f"[PositionTitleValidator] Error processing position title: {e}")
            return False

        return self.pass_to_next_validator(job)

File: job-scraper-service/internal/validator/position_title_validator.go
package validator

import (
	"regexp"
	"strings"

	"job-scraper-service/internal/models"
)

type PositionTitleValidator struct {
	BaseValidator
	roleRegex      *regexp.Regexp
	seniorityRegex *regexp.Regexp
}

func NewPositionTitleValidator() *PositionTitleValidator {
	rolePatterns := []string{
		`\bsoftware engineer\b`, `\bsoftware developer\b`, `\bsde\b`,
		`\bbackend engineer\b`, `\bback[- ]end engineer\b`,
		`\bfull[- ]stack engineer\b`, `\bapplication engineer\b`,
		`\bmachine learning engineer\b`, `\bml engineer\b`, `\bmle\b`,
		`\bdata scientist\b`, `\bdata science\b`, `\bapplied scientist\b`,
		`\bai engineer\b`, `\bgenai engineer\b`, `\bllm engineer\b`,
	}

	seniorityPatterns := []string{
		`\bsenior\b`, `\bsr\.?\b`, `\blead\b`, `\bstaff\b`, `\bprincipal\b`,
		`\bmanager\b`, `\bdirector\b`, `\barchitect\b`,
	}

	return &PositionTitleValidator{
		BaseValidator:  NewBaseValidator(),
		roleRegex:      regexp.MustCompile("(?i)" + strings.Join(rolePatterns, "|")),
		seniorityRegex: regexp.MustCompile("(?i)" + strings.Join(seniorityPatterns, "|")),
	}
}

func (p *PositionTitleValidator) Validate(job models.ScrapedJob) bool {
	if job.Title == "" {
		p.RecordStat(job.Platform, "fail")
		return false
	}

	// 1. Exclude high-level management and seniority items
	if p.seniorityRegex.MatchString(job.Title) {
		p.RecordStat(job.Platform, "fail")
		return false
	}

	// 2. Validate structural target matches
	if !p.roleRegex.MatchString(job.Title) {
		p.RecordStat(job.Platform, "fail")
		return false
	}

	p.RecordStat(job.Platform, "pass")
	return p.PassToNext(job)
}
